Deserialize publish requests on generic thread-pool

nicktindall opened the pull request:
This PR moves the `publish_state` handler from the `CLUSTER_COORDINATION` thread pool to the `GENERIC` one. This means the initial handling of the publish request, including the deserialisation of the cluster state, happens on one of the `GENERIC` threads instead of the `CLUSTER_COORDINATION` thread. Once we have deserialised the cluster state and done some validation, we delegate to the `CLUSTER_COORDINATION` pool to apply the new state.

The consequences of this include
- The generic pool contains multiple threads, but the cluster coordination pool contains a single thread. So (in theory) multiple instances of the handler could now be executing concurrently, where previously these messages would be handled serially.
- The delegation to the cluster coordination thread for the apply means we add some more asynchrony.

Closes #106352

nicktindall committed:
Deserialize publish requests on generic thread-pool

Closes #106352

nicktindall committed:
Update docs/changelog/108814.yaml

elasticsearchmachine commented:
Hi @nicktindall, I've created a changelog YAML for you.

nicktindall commented on server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java:
```diff
@@ -197,20 +200,41 @@ private PublishWithJoinResponse handleIncomingPublishRequest(BytesTransportReque
+            public void onResponse(PublishWithJoinResponse publishWithJoinResponse) {
+                onSuccess.accept(incomingState);
+                transportChannel.sendResponse(publishWithJoinResponse);
+            }
```
I considered using `ActionListener.runBefore(..., () -> onSuccess.apply(initialState))` but I think just putting the call to `onSuccess` inline in `onResponse` is a bit clearer (it's easy to miss the `runBefore` wrapping IMO, but I'm still new to these utilities)
I know using `runBefore` would behave better in the event onSuccess threw an exception, but it's only setting an `AtomicReference` so I don't think that's even possible the way things are.
Open to feedback on that.

DaveCTurner replied:
If we passed in an `ActionListener` rather than a `TransportChannel` then the `runBefore` would not be so lost in the noise I think.

nicktindall commented on server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java:
```diff
@@ -197,20 +200,41 @@ private PublishWithJoinResponse handleIncomingPublishRequest(BytesTransportReque
+                    acceptState(
+                        incomingState,
+                        transportChannel,
+                        (acceptedState) -> lastSeenClusterState.compareAndSet(lastSeen, acceptedState)
```
`onSuccess` will end up being called by the `CLUSTER_COORDINATION` thread only after a successful apply, so I wonder if we need to use compare-and-set here, I don't _think_ it's possible for `lastSeenCluster` to have changed between this being dispatched AND the new state being successfully applied. I would think either
- the state _has not_ changed and `onSuccess` is called correctly
- the state _has_ changed and the apply fails due to the version check, so `onSuccess` is not called

nicktindall replied:
The reason I raise this is because we could do away with the `onSuccess` callback if we were able to safely call `lastSeenClusterState.set` for both the full and diff payloads.

DaveCTurner replied:
I think it can change in between, dispatch and execution, yes - there could be another update in flight when we read the value when then completes and updates it before we get to run.

I am however not sure whether this matters. I'll think about this a little more.

nicktindall replied:
This one still concerns me a bit... the task of applying an update is done in three steps
1. deserialise payload
2. apply payload if valid
3. send response

only (2) happens in the mutex. So you could have the following interleaving for payload (a - version 7, b - version 8)

1. a-1
2. b-1
3. a-2 (succeeds, bump version to 7)
4. b-2 (successds, bump version to 8)
5. b-3 (set lastSeen to 8)
6. c-3 (set lastSeen to 7) - this means the next diff would be applied to 7, and a 7/9 hybrid would be applied locally?

The above could only happen if there were multiple threads in the `CLUSTER_COORDINATION` pool, but you said that is configurable.

We could put an additional check to only bump lastSeen when term and version are >= existing term and version?

Do we run tests with `CLUSTER_COORDINATION` size > 1 ?

DaveCTurner replied:
The sequence of 6 steps is indeed something that could have happened prior to #83576 and can again happen after this change - we could end up having applied state version 8 but with `lastSeen` at state version 7. But that's ok, each diff includes a UUID which identifies the base version which we check here:

https://github.com/elastic/elasticsearch/blob/9ba5651e7478216f17d81b63385ddf6db2331f7f/server/src/main/java/org/elasticsearch/cluster/ClusterState.java#L1170-L1172

That `IncompatibleClusterStateVersionException` is caught and handled here:

https://github.com/elastic/elasticsearch/blob/638a45009cb76059dd914c70eb4aa7e777768da2/server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java#L473-L483

IOW if we receive a diff between versions 8 & 9 but `lastSeen` is at version 7 then we reject the diff and the master sends us the full state at version 9 for us to apply. Somewhat inefficient for sure but still correct.

This case will be exercised in `org.elasticsearch.cluster.coordination.CoordinatorTests` in a `cluster.runRandomly()` call, but with rather low probability I think. We could try adding a test there which specifically checks this case.

nicktindall replied:
Great! not a problem we need to solve here then :)

nicktindall committed:
Increase AbstractCoordinatorTestCase
#CLUSTER_STATE_UPDATE_NUMBER_OF_DELAYS

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java:
```diff
@@ -197,20 +200,41 @@ private PublishWithJoinResponse handleIncomingPublishRequest(BytesTransportReque
+     * @param transportChannel The channel on which it was received
+     * @param onSuccess The action to perform if the new state is accepted
+     */
+    private void acceptState(ClusterState incomingState, TransportChannel transportChannel, Consumer<ClusterState> onSuccess) {
```
I think I'd rather we avoided passing the `TransportChannel` directly here. It works, but it's more idiomatic to have async methods accept an `ActionListener<...>` - see the Javadocs for `org.elasticsearch.action.ActionListener` for lots more detail, and also it's useful to constrain the response type as early as possible. In this case a `org.elasticsearch.action.support.ChannelActionListener<PublishWithJoinResponse>` would be a good choice. I think the same comment applies to `handleIncomingPublishRequest` - on the one hand, using a method reference here is kinda neat, but on the other hand we don't use the `Task` argument and it would be easier on the reader to pass in an `ActionListener<PublishWithJoinResponse>` here too.

nicktindall replied:
Yes that looks much nicer, thanks!

DaveCTurner reviewed (commented):
Some initial comments, mostly stylistic tho.

nicktindall committed:
Use ActionListeners to tidy up async calls

nicktindall committed:
Refactor to tidy
We can make incomingState final if we factor out the deserialisation
and application of the diff

nicktindall committed:
Rename listener parameter

nicktindall commented on server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java:
```diff
@@ -145,72 +151,88 @@ private PublishWithJoinResponse handleIncomingPublishRequest(BytesTransportReque
-                        assert false : e;
-                        throw e;
-                    }
+                    final ClusterState incomingState = deserializeAndApplyDiff(request, in, lastSeen);
```
This is probably controversial, but by factoring out the `deserialiseAndApplyDiff`, incomingState can be made final and used directly in the runBefore lambda. Git also appears to retain the history for the body of `deserialiseAndApply` nicely.

DaveCTurner replied:
Nope seems fine to me, nice idea.

nicktindall committed:
Only set lastSeen when it was successfully applied

nicktindall committed:
Restore compare-and-set

DaveCTurner commented on test/framework/src/main/java/org/elasticsearch/cluster/coordination/AbstractCoordinatorTestCase.java:
```diff
@@ -219,11 +219,12 @@ protected static int defaultInt(Setting<Integer> setting) {
-    // 8. apply committed state on master (last one to apply cluster state)
-    // 9. complete the publication listener back on the master service thread
-    public static final int CLUSTER_STATE_UPDATE_NUMBER_OF_DELAYS = 9;
+    // 6. nodes deserialize committed cluster state
+    // 7. nodes apply committed cluster state
+    // 8. master receives ApplyCommitResponses
+    // 9. apply committed state on master (last one to apply cluster state)
+    // 10. complete the publication listener back on the master service thread
+    public static final int CLUSTER_STATE_UPDATE_NUMBER_OF_DELAYS = 10;
```
:+1: except that this extra step is happening between steps 3 & 4 in the old list. `org.elasticsearch.cluster.coordination.ApplyCommitRequest` only carries the term and version, it's `org.elasticsearch.cluster.coordination.PublishRequest` which carries the state that's being published.

DaveCTurner reviewed (commented):
one nit otherwise LGTM

nicktindall committed:
Fix NUMBER_OF_DELAYS description

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java:
```diff
@@ -145,72 +151,88 @@ private PublishWithJoinResponse handleIncomingPublishRequest(BytesTransportReque
-                    final PublishWithJoinResponse response = acceptState(incomingState);
-                    lastSeenClusterState.compareAndSet(lastSeen, incomingState);
-                    return response;
+                    acceptState(incomingState, publishResponseListener.delegateFailure((delegate, response) -> {
+                        lastSeenClusterState.compareAndSet(lastSeen, incomingState);
+                        delegate.onResponse(response);
+                    }));
```
Sorry one more tiny thing, I would write this sort of thing using `ActionListener#map` to avoid any doubt that we complete the listener on all paths:

```suggestion
                    acceptState(incomingState, publishResponseListener.map(response -> {
                        lastSeenClusterState.compareAndSet(lastSeen, incomingState);
                        return response;
                    }));
```

Not a huge deal here ofc but a useful trick in more complex cases for sure.

nicktindall replied:
Hehe yeah it took me a while to work out which helper to use here. Will make the change.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java:
```diff
@@ -145,72 +151,88 @@ private PublishWithJoinResponse handleIncomingPublishRequest(BytesTransportReque
-                final PublishWithJoinResponse response = acceptState(incomingState);
-                lastSeenClusterState.set(incomingState);
-                return response;
+                acceptState(incomingState, publishResponseListener.delegateFailure((delegate, response) -> {
+                    lastSeenClusterState.set(incomingState);
+                    delegate.onResponse(response);
+                }));
```
Likewise here.

```suggestion
                acceptState(incomingState, publishResponseListener.map(response -> {
                    lastSeenClusterState.set(incomingState);
                    return response;
                }));
```

DaveCTurner reviewed (approved):
LGTM (one more tiny suggestion but no need for another review)

nicktindall committed:
Update server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java

Co-authored-by: David Turner <david.turner@elastic.co>

nicktindall committed:
Update server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java

Co-authored-by: David Turner <david.turner@elastic.co>
